### PR TITLE
creating memory game

### PR DESCRIPTION
**Pull request by emilymaskin**

### CODE REVIEW

File: js/memory.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+					var thisCard = deck.cards.pop();
+					$boardEl.append(thisCard.createEl());
+				}
+				$boardEl.append('<br>');
+			}
+		};
+
+		var memoryBoard = new Board(4);
+		memoryBoard.dealBoard();
+	})();
```
Comment (afeld) [inline js/memory.js:142]: Do you actually need this self-executing function?

File: js/memory.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+});
```
Comment (afeld) [inline js/memory.js:143]: Note that the class definitions don't actually need to be within the `document.ready()`...usually it's preferable to just wrap the method call that will actually require it.

File: js/memory.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
+$(document).ready(function() {
+
+	(function() {
+		'use strict';
+
+		var cardAlreadyShowing = null;
+		var totalTries = 0;
+		var $boardEl = $('#board');
```
Comment (afeld) [inline js/memory.js:8]: Rather than making them nearly-global variables, where else could you put these?

File: js/memory.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+				.addClass('each-card')
+				.data('background-color', this.color)
+				.on('click', function () {
+					thisCard.showCard();
+					if (cardAlreadyShowing) {
+						thisCard.checkForMatch(thisCard);
+					} else {
+						cardAlreadyShowing = thisCard;
+					}
+				});
```
Comment (afeld) [inline js/memory.js:32]: You might want to move the click handler to be another method on the `Card`, just to keep `createEl()` short. I've even heard people go so far as to say "don't use anonymous functions for callbacks".

File: js/memory.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+			this.cardEl.css('background-color', '#ccc');
+		};
+
+		Card.prototype.removeFromPlay = function() {
+			this.cardEl.addClass('out-of-play');
+		};
+
+		Card.prototype.checkForMatch = function(thisCard) {
+			$boardEl.addClass('checking'); // disable clicks while processing
+			if (thisCard !== cardAlreadyShowing && thisCard.color === cardAlreadyShowing.color) {
```
Comment (afeld) [inline js/memory.js:50]: How about an `isMatch()` method?

File: js/memory.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+				/* if they clicked on two (different) cards of the same color */
+				this.handleMatch(thisCard);
+			} else {
+				this.handleMismatch(thisCard);
+			}
+			totalTries++;
+			if ($('.out-of-play').length === $('.each-card').length) {
```
Comment (afeld) [inline js/memory.js:57]: What is this line doing? (tell me via a descriptive method name :wink:)

File: js/memory.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+		Card.prototype.handleMismatch = function(thisCard) {
+			/* show colors for one second, then hide again */
+			setTimeout(function() {
+				thisCard.hideCard();
+				cardAlreadyShowing.hideCard();
+				cardAlreadyShowing = null;
+				$boardEl.removeClass('checking');
+			}, 1000);
+		};
```
Comment (afeld) [inline js/memory.js:78]: To keep them simple, you might consider having the Cards only care about themselves (showing, hiding, etc.), and have the containing class (the Board) worry about checking for and handling matches.

File: js/memory.js [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+					color = this.getNextColor();
+				}
+				this.cards.push(new Card(color));
+			}
+		};
+
+		Deck.prototype.shuffle = function() {
+			this.cards = this.cards.sort(function() {
+				return 0.5 - Math.random();
+			});
```
Comment (afeld) [inline js/memory.js:105]: Nice! Was this your idea, or did you find the trick somewhere?

File: js/memory.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+		};
+
+		Deck.prototype.getNextColor = function() {
+			var redVal = Math.floor(Math.random() * 256);
+			var greenVal = Math.floor(Math.random() * 256);
+			var blueVal = Math.floor(Math.random() * 256);
+			return 'rgba(' + redVal + ',' + greenVal + ',' + blueVal + ', 0.7)';
```
Comment (afeld) [inline js/memory.js:112]: Ditto.

File: js/memory.js [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+		/**
+		 * Board constructor
+		 *
+		 * @constructor
+		 * @param {number} size The number of cards on one side of the square
+		 */
+		var Board = function(size) {
+			this.size = size;
+			this.cardsShown = [];
+			this.totalCardCount = this.size * this.size;
```
Comment (afeld) [inline js/memory.js:124]: In general, I wouldn't store a property that can be re-computed from other data. In other words, I wouldn't store the `size` _and_ the `totalCardCount`. Make one of them a method.

**Comment by afeld** [comment]
Really nice organization of code, simple methods, etc...great job!

File: js/memory.js (hunk 8 above)
Comment (emilymaskin) [inline js/memory.js:105]: The one above I googled; the one below I came up with :)

Comment (afeld) [inline js/memory.js:105]: Per [the syllabus](http://documentup.com/advanced-js/syllabus#statements-on-plagiarism/instructor), please link to the original!

Comment (emilymaskin) [inline js/memory.js:105]: Oops! Done

**Pull request closed without merging**